### ISSUE TITLE
generalize not failing when match fails

### ISSUE THREAD

**Issue by tobiasgrosser** [leanprover/lean4#3889]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

I am trying to write a tactic that generalizes patterns in a program. My tactic calls `generalize_or_fail` in `repeat` until all occurrences of the pattern are generalized. I cannot just use `generalize` as it does not fail, but just keeps creating meta-variables in case the match pattern cannot be found. I managed to create a workaround, as shown below. However, this workaround is also
not working for us as `conv` seems to solve goals by reflection, as visible in `test_solved_but_should_not`. In this specific case, this leads to `generalize` failing, which I can work around with `try`. While this resolves the error it leads to a solved goal state
while I just wanted to generalize. In our framework, I would much prefer if our tactic does not start solving the goal.

```lean
local macro "generalize_or_fail" : tactic =>
  `(tactic|
   (
       conv in ((_ : ℕ) + (_ : ℕ)) => skip
       generalize h : ((_ : ℕ) + (_ : ℕ)) = c
    )
  )

def test_succeed (a b c : ℕ): a + b + c = c + (a + b) := by
  generalize_or_fail
  -- a b c c✝ : ℕ
  -- h✝ : a + b + c = c✝
  -- ⊢ c✝ = c + (a + b)
  sorry

def test_fail (a b c : ℕ): a * b * c = c * (a * b) := by
  generalize_or_fail
  -- 'pattern' conv tactic failed, pattern was not found
  -- fun x_0 x_1 => x_0 + x_1
  sorry

def test_solved_but_should_not (a b c : ℕ): a + b = a + b := by
  generalize_or_fail
  -- no goals to be solved
```

### Steps to Reproduce

Run the code above.

**Expected behavior:** 
  - generalize should fail if it cannot match a pattern
or  
  - conv should not simplify by refl
**Actual behavior:**
  - generalize introduces a new meta-variable even if no pattern is found
  - conv simplifies by reflection

### Versions

"4.7.0"

**Comment by tobiasgrosser** [comment]
There is a corresponding zulip thread where we investigate this further: https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/conv.20tactic.20sometimes.20solves.20goal/near/432918099

**Comment by tobiasgrosser** [comment]
I meanwhile use the following tactic:

```lean
elab "contains? " ts:term : tactic => withMainContext do
  let tgt ← getMainTarget
  if (← kabstract tgt (← elabTerm ts none)) == tgt then throwError "pattern not found"
```

which does what I need. Thanks goes to Kyle Miller. 

I am happy for this to be closed, but also wanted to record this in case someone feels these issues should be addressed within lean core.

**Comment by semorrison** [comment]
I agree `generalize` should fail if the pattern doesn't match. If someone would like to PR a fix, that would be welcome.

**Comment by samhithak13** [comment]
I'd like to take this one.

Proposed scoping: `generalize` should keep succeeding when the term contains no placeholders and simply doesn't occur in the goal and that's the existing "just add equality" idiom, and there's a test relying on it in `tests/elab/generalize.lean`. The error would fire only when the term contains placeholders (`_`) that are still unassigned after `kabstract` has attempted to match, since those are genuinely undetermined and currently leak metavariables into the new hypothesis.

I'd put the check in the `generalize` tactic elaborator rather than in `Lean.MVarId.generalize`, since the latter is also called with user-written terms by `cases h : e`, `induction`, and `split`, where the same restriction shouldn't apply.

Does that scoping sound right?
